File: examples/n_queens.rs
//! N-Queens: place N queens on an N×N board so that no two attack each other.
//!
//! A permutation genome puts one queen in each row and each column (queen `row` is in column
//! `order[row]`), so only the diagonals can conflict. Permutations have no position-wise
//! crossover, so this uses (μ+λ) with swap mutation only.
//!
//! ```text
//! cargo run --release --example n_queens
//! ```

use genoxide::prelude::*;

const N: usize = 64;

// the number of pairs of queens on the same diagonal
fn conflicts(order: &Order) -> f64 {
    let mut diagonals = vec![0usize; 2 * N];
    let mut anti_diagonals = vec![0usize; 2 * N];
    for (row, &column) in order.iter().enumerate() {
        diagonals[row + N - column] += 1;
        anti_diagonals[row + column] += 1;
    }
    let pairs = |count: &usize| count * count.saturating_sub(1) / 2;
    (diagonals.iter().map(pairs).sum::<usize>() + anti_diagonals.iter().map(pairs).sum::<usize>())
        as f64
}

fn main() -> Result<()> {
    let ga = Ga::builder(Permutation::new(N)?)
        .population_size(20)
        .select(Tournament::new(2)?)
        .crossover(NoCrossover)
        .mutate(SwapMutation::new())
        .scheme(Scheme::MuPlusLambda { lambda: 20 })
        .minimize()
        .seed(1)
        .build()?;

    let outcome = Engine::new(ga, conflicts)
        .stop_when(Stop::target(0.0).or(Stop::generations(50_000)))
        .run()?;

    println!(
        "{:?} after {} generations: {} conflicts",
        outcome.stop_reason(),
        outcome.generations(),
        outcome.best_fitness()
    );
    if N <= 16 {
        for &column in outcome.best_genome().iter() {
            let row: String = (0..N)
                .map(|c| if c == column { 'Q' } else { '.' })
                .collect();
            println!("{row}");
        }
    }
    Ok(())
}

File: python/examples/n_queens.py
"""N-Queens with 64 queens: a permutation puts one queen per row and per column, and tabu search
removes the diagonal conflicts."""

import numpy as np

import genoxide as gx

n = 64
rows = np.arange(n)


def conflicts(order):
    """Queens that share a diagonal with another queen."""
    return float(2 * n - len(np.unique(order - rows)) - len(np.unique(order + rows)))


search = gx.LocalSearch(
    gx.Permutation(n),
    neighbor=gx.SwapMutation(),
    neighbors=32,
    acceptance=gx.Tabu(20),
    objective="minimize",
    seed=1,
)
result = search.run(conflicts, target=0, evaluations=1_000_000)
print(f"{result.best_fitness:.0f} conflicts after {result.evaluations} evaluations")
